### PR TITLE
track heartbeat, login and restores for mobile for replay attack

### PR DESCRIPTION
**Pull request by mkangia**
[Details](https://docs.google.com/document/d/1O4SD_vNBNzvnmy4YsrwsO1VVQA-9ur9LwB0RyG6yJSI/edit)

1. Check for replay attack for mobile requests for login, heartbeat and restore [done]
2. improve logic to expire obfuscated passwords older than 3 days [done]

Test:
1. behaviour in case of 401 requests. Are their username present in case of the initial 401's sent by phone?

QA:
Web
1. login in successfully by a mobile user
2. login in with 2-factor auth by a web user
3. on replay attack fails login by a web user

Mobile:
1. login in successfully
2. does not track for user not asked to track for
3. does not track request not to be tracked like form submission
4. test replay attack blocking/returning 401 for requests to be tracked
5. test replay attack not blocking for requests not to be tracked

### CODE REVIEW

File: custom/nic_compliance/utils.py [hunk 1]
```diff
@@ -90,11 +95,15 @@ def record_login_attempt():
         client.set(key_name, obfuscated_passwords + [hash_password(obfuscated_password)])
         client.expire(key_name, timedelta(EXPIRE_LOGIN_ATTEMPTS_IN))
 
+    def _mobile_request_to_track():
+        return resolve(request.path).url_name in MOBILE_REQUESTS_TO_TRACK_FOR_REPLAY_ATTACK
```
Comment (mkangia) [inline custom/nic_compliance/utils.py]: To be able to test this without effecting resources at scale, i would like to test this by adding a check for username like
`username == 'test-nic-compliance'`?
Does that sound okay or is there a better way?
I can't use FF or domain specific check here.

**Comment by mkangia** [comment]
@snopoke 
there is a pending change to improve password expiry but an early review on the new implementation would be helpful.

File: custom/nic_compliance/utils.py (hunk 1 above)
Comment (snopoke) [reply custom/nic_compliance/utils.py]: Fine to hardcode a username for a test.

File: custom/nic_compliance/utils.py [hunk 2]
```diff
@@ -66,35 +72,41 @@ def verify_password(password, encoded_password):
...
-    return REDIS_LOGIN_ATTEMPTS_LIST_PREFIX + username
+def obfuscated_password_redis_key_for_user(username, obfuscated_password):
+    return '_'.join([REDIS_LOGIN_ATTEMPTS_LIST_PREFIX, username, obfuscated_password])
 
 
-def get_obfuscated_passwords(username):
+def get_obfuscated_passwords(username, obfuscated_password):
     client = get_redis_client()
-    return client.get(obfuscated_passwords_redis_key_for_user(username), [])
+    return client.get(obfuscated_password_redis_key_for_user(username, obfuscated_password), [])
```
Comment (snopoke) [inline custom/nic_compliance/utils.py]: should be `client.get(...., False)`

File: custom/nic_compliance/const.py [hunk 3]
```diff
@@ -1,6 +1,18 @@
 from __future__ import unicode_literals
 REDIS_USED_PASSWORDS_LIST_PREFIX = 'used_passwords_'
-REDIS_LOGIN_ATTEMPTS_LIST_PREFIX = 'login_attempts_'
+REDIS_LOGIN_ATTEMPTS_LIST_PREFIX = 'login_attempts'
```
Comment (snopoke) [inline custom/nic_compliance/const.py]: can we make this smaller to use less memory?

File: custom/nic_compliance/tests/test_utils.py [hunk 4]
```diff
@@ -33,8 +42,10 @@ def tearDownClass(cls):
         cls.web_user.delete()
         super(TestDecodePassword, cls).tearDownClass()
 
+    @patch("custom.nic_compliance.utils.USERS_TO_TRACK_FOR_REPLAY_ATTACK", [username])
     def test_login_attempt(self):
```
Comment (snopoke) [inline custom/nic_compliance/tests/test_utils.py]: should add a test for one of the new endpoints

File: custom/nic_compliance/utils.py (hunk 2 above)
Comment (mkangia) [reply custom/nic_compliance/utils.py]: Dont need the method anymore, removing it.

File: custom/nic_compliance/utils.py [hunk 5]
```diff
@@ -66,35 +72,36 @@ def verify_password(password, encoded_password):
     return PASSWORD_HASHER.verify(password, encoded_password)
 
 
-def obfuscated_passwords_redis_key_for_user(username):
-    return REDIS_LOGIN_ATTEMPTS_LIST_PREFIX + username
+def obfuscated_password_redis_key_for_user(username, obfuscated_password):
+    return '_'.join([REDIS_LOGIN_ATTEMPTS_LIST_PREFIX, username, obfuscated_password])
```
Comment (snopoke) [inline custom/nic_compliance/utils.py]: can we make this prefix shorter to reduce memory requirement

Comment (mkangia) [reply custom/nic_compliance/utils.py]: i updated `REDIS_LOGIN_ATTEMPTS_LIST_PREFIX` to be just `la$$`. Rest is username and obfuscated password.

We need username and obfuscated password to identify this uniquely. Can also do md5 hash of the username & password and use that as the key instead

for example:
```
In [16]: username
Out[16]: 'mkangia@dimagi.com'

In [17]: obf_password
Out[17]: 'sha1$SUXeJGr1jWhM$902d24b15a1cb63f5b6c989d53517ed76764c7ef'

In [18]: REDIS_LOGIN_ATTEMPTS_LIST_PREFIX  + hashlib.md5("%s%s"% (username, obf_password)).hexdigest()
Out[18]: '$$la901270bcbe77d5ce914c60592a201d38'
```

**Review by snopoke — APPROVED** [review]

**Pull request merged**